crates/rooch-framework/sources/auth_validator/ecdsa_k1_validator.move

```diff
@@ -83,61 +82,95 @@ module rooch_framework::ecdsa_k1_validator {
         sign
     }
 
-    /// Get the authentication key of the given authenticator.
-    public fun ecdsa_k1_authentication_key(authenticator_payload: &vector<u8>): vector<u8> {
-        let public_key = ecdsa_k1_public_key(authenticator_payload);
-        let addr = ecdsa_k1_public_key_to_address(public_key);
+    /// Get the authentication key of the given authenticator from authenticator_payload.
+    public fun get_authentication_key_from_authenticator_payload(authenticator_payload: &vector<u8>): vector<u8> {
+        let public_key = get_public_key_from_authenticator_payload(authenticator_payload);
+        let addr = public_key_to_address(public_key);
         moveos_std::bcs::to_bytes(&addr)
     }
 
-    public fun ecdsa_k1_public_key_to_address(public_key: vector<u8>): address {
+    public fun public_key_to_address(public_key: vector<u8>): address {
+        moveos_std::bcs::to_address(public_key_to_authentication_key(public_key))
```

Comment:
ecdsa_k1 do not support public key to address. Or do we support generating the ETH address via the ecdsa_k1 public key?

Reply:
Yes. Both ecdsa_k1 and ecdsa_k1_recoverable's public key length are 33 bytes. The public key can then be used to generate ETH addresses:

```move
let uncompressed = decompress_pubkey(&pubkey);

// Take the last 64 bytes of the uncompressed pubkey.
let uncompressed_64 = vector::empty<u8>();
let i = 1;
while (i < 65) {
 let value = vector::borrow(&uncompressed, i);
 vector::push_back(&mut uncompressed_64, *value);
 i = i + 1;
};

// Take the last 20 bytes of the hash of the 64-bytes uncompressed pubkey.
let hashed = hash::keccak256(&uncompressed_64);
let addr = vector::empty<u8>();
let i = 12;
while (i < 32) {
 let value = vector::borrow(&hashed, i);
 vector::push_back(&mut addr, *value);
 i = i + 1;
};

addr
```

I think the same way applies for Bitcoin address using ripemd-160 hash.

Reply:
May split `public_key_to_address` to `public_key_to_ethereum_address` and `public_key_to_bitcoin_address`. Just need to determine which address to use as the authentication key.

Reply:
So the address should not be `address`. We should define an EVMAddress type.

Reply:
```rs
impl AccountAddress {
    pub const fn new(address: [u8; Self::LENGTH]) -> Self {
        Self(address)
    }

    /// The number of bytes in an address.
    /// Default to 16 bytes, can be set to 20 bytes with address20 feature.
    pub const LENGTH: usize = if cfg!(feature = "address20") {
        20
    } else if cfg!(feature = "address32") {
        32
    } else {
        16
    };
```

Both ETH and BTC are in 20-byte address. While although Rooch address is in 32-byte address, according to the above code from `MoveValue::Address`, it can be set to 20 bytes if needed? Or in anyway EVMAddress type is what we need and we should define it in `move_core_types::value::MoveValue` or somewhere else?

Reply:
A handling way to implement `address20` feature is:

```toml
move-core-types = { git = "https://github.com/rooch-network/move", rev = "55b8bf6967e76ff11933fffaccb76d5a4e15a23a", features = ["address32", "address20"] }
```

But it needs to be well tested in all imported libraries and could be a breaking change.

Reply:
We need to define the EVMAddress in Move. The `address` type is only for Rooch or Move addresses.

Reply:
Like https://github.com/starcoinorg/starcoin-framework/blob/1f8ebec924fe8768888b02975614650c4a1c63c9/sources/Signature.move#L46